public/app/features/alerting/unified/components/rules/RuleListStateView.tsx

```diff
@@ -32,34 +40,97 @@ export const RuleListStateView = ({ namespaces }: Props) => {
           // In this case, we shouldn't try to group these alerts in the state view
           // Even though we handle this at the API layer, this is a last catch point for any edge cases
           if (rule.promRule && isAlertingRule(rule.promRule) && rule.promRule.state) {
-            result[rule.promRule.state].push(rule);
+            result.get(rule.promRule.state)?.push(rule);
           }
         })
       )
     );
 
-    Object.values(result).forEach((rules) => rules.sort((a, b) => a.name.localeCompare(b.name)));
+    result.forEach((rules) => rules.sort((a, b) => a.name.localeCompare(b.name)));
 
     return result;
   }, [namespaces]);
+
+  const entries = groupedRules.entries();
+
   return (
-    <>
-      {(!filters.alertState || filters.alertState === PromAlertingRuleState.Firing) && (
-        <RuleListStateSection state={PromAlertingRuleState.Firing} rules={groupedRules[PromAlertingRuleState.Firing]} />
-      )}
-      {(!filters.alertState || filters.alertState === PromAlertingRuleState.Pending) && (
-        <RuleListStateSection
-          state={PromAlertingRuleState.Pending}
-          rules={groupedRules[PromAlertingRuleState.Pending]}
-        />
-      )}
-      {(!filters.alertState || filters.alertState === PromAlertingRuleState.Inactive) && (
-        <RuleListStateSection
-          defaultCollapsed={filters.alertState !== PromAlertingRuleState.Inactive}
-          state={PromAlertingRuleState.Inactive}
-          rules={groupedRules[PromAlertingRuleState.Inactive]}
+    <ul className={styles.columnStack} role="tree">
+      {Array.from(entries).map(([state, rules]) => (
+        <RulesByState key={state} state={state} rules={rules} />
+      ))}
+    </ul>
+  );
+};
+
+const STATE_TITLES: Record<PromAlertingRuleState, string> = {
+  [PromAlertingRuleState.Firing]: 'Firing',
+  [PromAlertingRuleState.Pending]: 'Pending',
+  [PromAlertingRuleState.Inactive]: 'Normal',
+};
+
+const RulesByState = ({ state, rules }: { state: PromAlertingRuleState; rules: CombinedRule[] }) => {
+  const { page, pageItems, numberOfPages, onPageChange } = usePagination(rules, 1, DEFAULT_PER_PAGE_PAGINATION);
+
+  const isFiringState = state !== PromAlertingRuleState.Firing;
+  const hasRulesMatchingState = rules.length > 0;
+
+  return (
+    <ListSection
+      title={
+        <Stack alignItems="center" gap={0}>
+          {STATE_TITLES[state] ?? 'Unknown'}
+          <Counter value={rules.length} />
+        </Stack>
+      }
+      collapsed={isFiringState || hasRulesMatchingState}
+      pagination={
+        <Pagination
+          currentPage={page}
+          numberOfPages={numberOfPages}
+          onNavigate={onPageChange}
+          hideWhenSinglePage={true}
         />
-      )}
-    </>
+      }
+    >
+      {pageItems.map((rule) => {
+        const { rulerRule, promRule } = rule;
+
+        const isProvisioned = isGrafanaRulerRule(rulerRule) && Boolean(rulerRule.grafana_alert.provenance);
```

Comment:
not for this PR, but we could add the `isProvisionedRulerRule` function in rules.ts?

Reply:
Great idea! I'll append that 👍